kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateCloudRepository.java

```diff
@@ -0,0 +1,112 @@
+
+import com.thoughtworks.xstream.XStream;
+import com.thoughtworks.xstream.converters.reflection.PureJavaReflectionProvider;
+import io.fabric8.kubernetes.api.model.ConfigMap;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import org.kie.server.api.KieServerConstants;
+import org.kie.server.api.model.KieContainerResource;
+import org.kie.server.api.model.KieServerConfigItem;
+import org.kie.server.services.impl.storage.KieServerState;
+import org.kie.server.services.impl.storage.KieServerStateRepository;
+import org.kie.server.services.openshift.api.KieServerReadinessProbe;
+import org.kie.soup.commons.xstream.XStreamUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KieServerStateCloudRepository implements KieServerStateRepository,
+                                           KieServerReadinessProbe, CloudClientFactory {
+
+    public static final String ROLLOUT_REQUIRED = "services.server.kie.org/openshift-startup-strategy.rolloutRequired";
+    public static final String CFG_MAP_DATA_KEY = "kie-server-state";
+    public static final String CFG_MAP_LABEL_NAME = "services.server.kie.org/kie-server-state";
+    public static final String CFG_MAP_LABEL_VALUE = "USED";
+    
+    protected static final String STATE_CHANGE_TIMESTAMP = "services.server.kie.org/kie-server-state.changeTimestamp";
+    private static final Logger logger = LoggerFactory.getLogger(KieServerStateCloudRepository.class);
+
+    protected final XStream xs;
+    
+    public static XStream initializeXStream() {
+        XStream xs = XStreamUtils.createTrustingXStream(new PureJavaReflectionProvider());
+        String[] voidDeny = {"void.class", "Void.class"};
+        xs.denyTypes(voidDeny);
+        xs.alias(CFG_MAP_DATA_KEY, KieServerState.class);
+        xs.alias("container", KieContainerResource.class);
+        xs.alias("config-item", KieServerConfigItem.class);
+
+        return xs;
+    }
+
+    public KieServerStateCloudRepository() {
+        xs = initializeXStream();
+    }
+
+    @Override
+    @NotNull
+    public KieServerState load(@NotNull String serverId) {
+        KieServerState kieServerState = processKieServerState(client -> {
+            ConfigMap cm = client.configMaps().withName(serverId).get();
+            if (cm == null) {
+                throw new IllegalStateException(("KieServerId: [" + serverId + "], not found. Please create an associated ConfigMap with configuration first."));
+            }
+            return (KieServerState) xs.fromXML(cm.getData().get(CFG_MAP_DATA_KEY));
+        });
+
+        if (kieServerState == null || !retrieveKieServerId(kieServerState).equals(serverId)) {
+            throw new IllegalStateException("Invalid KieServerId: " + serverId + ", or inconsistent state data.");
+        }
+
+        return kieServerState;
+    }
+
+    @Override
+    public void store(String serverId, KieServerState kieServerState) {
+        throw new UnsupportedOperationException();
```

Comment:
@rhtevan What about making KieServerStateCloudRepository abstract class? This way we can indicate that it shouldn't be used directly and by removing this store() method we can force all implementations to implement it.